repoparse.py

```diff
@@ -0,0 +1,294 @@
+from dataclasses import dataclass
+from typing import Dict, List
+
+import sys
+
+from dulwich.objects import Commit, Blob
+from dulwich.patch import patch_filename, is_binary, gen_diff_header, unified_diff
+from dulwich.repo import Repo
+from os import getcwd, path
+
+from os import makedirs, system
+from shutil import rmtree
+
+from unidiff import PatchSet, UnidiffParseError
+
+from shared import flush_progress_bar, progress, finish_progress, group, s_get_content, s_get_lines, unwrap_bytes_gen_to_str
+
+import multiprocessing
+
+from treesitter_stage import load_treesitter
+from tree_sitter import Language, Parser
+
+@dataclass(init=True)
+class ParsedCommitInfo:
+
+    # Only single parent is allowed
+    parent_sha: str
+    patch_set: Dict[str, str]
+    langs: Dict[str, str]
+    ids: Dict[str, str]
+
+
+@dataclass
+class ParsedCommits:
+
+    commit_data_by_sha: Dict[str, ParsedCommitInfo]
+
+    def __getitem__(self, item: str) -> ParsedCommitInfo:
+        return self.commit_data_by_sha.get(item)
+
+    def __setitem__(self, key: str, value: ParsedCommitInfo):
+        self.commit_data_by_sha[key] = value
+
+
+@dataclass
+class ParsedRepo:
+
+    commits_by_author: Dict[str, ParsedCommits]
+
+    def __getitem__(self, item: str):
+        return self.commits_by_author.get(item)
+
+    def __setitem__(self, key: str, value: ParsedCommits):
+        self.commits_by_author[key] = value
+
+def cleanup_diff(diff):
+    return '\n'.join(map(lambda s: s[1:] if s[0] == '+' else s, diff))
+
+def patch_set_to_json(repopath, patchSet):
+    dict_obj = {}
+    for patched_file in patchSet:
+        changes_list = []
+        dict_obj[path.join(repopath, patched_file.path)] = changes_list
+        for hunk in patched_file:
+            changes_list.extend(map(str, hunk.target_lines()))
+    
+    return dict_obj
+    
+
+def repo_part_to_json(obj):
+    # Python can't pattern match the argument to a method parameter
+    # by type if we need to pass a method into another function (?)
+    if isinstance(obj, ParsedCommitInfo):
```

Comment:
let's create method in each dataclass that will implement this logic (you can even inherit from some base class).
In other case this logic is quite hard to understand